### ISSUE TITLE
Threading not working as expected

### ISSUE THREAD

**Issue by Resonious** [stalwartlabs/jmap-server#34]
Hey there.

I've been playing around with Stalwart JMAP in my off time - really impressive stuff.

I imported some real emails into Stalwart, and the threading seemed a bit odd. Specifically, I noticed new threads being created for replies, and also a Gravatar registration email ended up inside a seemingly unrelated thread.

For the former error, I was able to reproduce with some very simple hand-written messages. Haven't gotten a simple repro for the latter yet.

```
From: sender@example.com
To: recipient@example.com
Message-ID: <msg1@example.com>
Subject: my thread

Hello, message 1
```

```
From: recipient@example.com
To: sender@example.com
Message-ID: <msg2@example.com>
In-Reply-To: <msg1@example.com>
Subject: my thread

Hello, message 2
```

```
From: sender@example.com
To: recipient@example.com
Message-ID: <msg3@example.com>
In-Reply-To: <msg2@example.com>
Subject: my thread

Hello, message 3
```

(full import script here: https://gist.github.com/Resonious/54bf0d9681edcd45ae582b06833912ea)

When I import the above 3 messages, they each end up in a different thread. Am I misunderstanding the threading algorithm, or are these supposed to be the same thread?
![image](https://user-images.githubusercontent.com/2793160/219683139-d3a9b824-2cb8-42ca-9f7a-ff9c51348e11.png)

**Comment by mdecimus** [comment]
Hi,

Threading should work as you expect, those three messages should be assigned to the same thread.
I am travelling at the moment so I can't check the code right now but could you verify if replacing the In-Reply-To header with References produces the same or different results?

Thanks.

**Comment by Resonious** [comment]
Thanks for the quick reply. I hope I'm not ruining your vacation or anything.

I swapped out In-Reply-To for References with a brand new set of message IDs. Got these new Email objects:

```json
{
  "id": "iaaaabn",
  "blobId": "bc0ccaqit7k7k1ujbnggqbhw0xfodh22as0scbx2s0navshmgap3q",
  "threadId": "c",
  "size": 116
}
{
  "id": "maaaabo",
  "blobId": "bhysxhguxq02is33ucz1sssib7i9n9jfm7bg0svhm0crjrhshdw7q",
  "threadId": "d",
  "size": 147
}
{
  "id": "qaaaabp",
  "blobId": "btngtl29x9e0ihzx2qekqstcj2v30e3mflwqmkfs1c0ldkxjzp0cq",
  "threadId": "e",
  "size": 147
}
```

Seems like pretty much the same results. And actually I think all of these thread IDs are not new. They already contain other messages from completely different senders and recipients. Could be the same phenomenon that caused my Gravatar registration mail to show up in an unrelated thread.

**Comment by Resonious** [comment]
I wrote some new test cases that seem to expose the problem. Can see them in this PR: https://github.com/stalwartlabs/jmap-server/pull/35

**Comment by mdecimus** [comment]
> Thanks for the quick reply. I hope I'm not ruining your vacation or anything.

No problem, I was returning from vacation :)

> Seems like pretty much the same results. And actually I think all of these thread IDs are not new. They already contain other messages from completely different senders and recipients. Could be the same phenomenon that caused my Gravatar registration mail to show up in an unrelated thread.

Thanks for checking this out and the PR. I am currently finishing the SMTP server and after that I am planning a big refactoring of the JMAP server which among other things will add add support for SQLite, PostgreSQL and other databases. 
I will look into this as soon as the SMTP server is released (less than one month hopefully) and will let you know.

**Comment by zvasilev** [comment]
Hi ,

I would introduce a potential edge case: there is no assurance that emails will be received in a specific order, as the remote mail transfer agent (MTA) may be processing a large mail queue or undergoing maintenance for a period of time. To illustrate, suppose an email thread contains five messages, but the third email is the last to be received. This situation could result in two separate threads, which highlights the need for a means of joining them.

**Comment by mdecimus** [comment]
> I would introduce a potential edge case: there is no assurance that emails will be received in a specific order, as the remote mail transfer agent (MTA) may be processing a large mail queue or undergoing maintenance for a period of time. To illustrate, suppose an email thread contains five messages, but the third email is the last to be received. This situation could result in two separate threads, which highlights the need for a means of joining them.

The test suite takes into account this scenario of messages belonging to same thread arriving in a different order. This issue seems to be something else.

**Comment by mdecimus** [comment]
This issue has now been fixed in Stalwart JMAP v0.3. Please retest and if the issue persists reopen the PR. Thanks!
